src/edit_distance/osa.gleam

```diff
@@ -0,0 +1,193 @@
+/// Compute the edit distance between two strings using the Restricted Damerau–Levenshtein, also
+/// called the [Optimal String Alignment (OSA) distance](https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance#Optimal_string_alignment_distance).
+///  
+/// The OSA distance counts operations: Insertions, deletions, substitutions, and limited 
+/// transpositions (only adjacent transpositions are allowed).
+/// 
+/// ## Examples
+/// 
+/// ```gleam
+/// > distance("gleam", "beam")
+/// 2
+/// ```
+/// 
+/// ```gleam
+/// > distance("cat", "cap")
+/// 1
+/// ```
+/// 
+/// ```gleam
+/// > distance("wibble", "wibbel")
+/// 2
+/// ```
+/// 
+pub fn distance(one: String, other: String) -> Int {
+  case one, other {
+    // Base case: Both strings are empty, so the distance is 0
+    "", "" -> 0
+
+    // Case: One string is empty, distance is the length of the other (insertion cost)
+    "", _ -> string.length(other)
+    _, "" -> string.length(one)
+
+    // Case: Strings are identical, no cost
+    one, other if one == other -> 0
+
+    // General case: Compute the Restricted Damerau-Levenshtein distance recursively
+    _, _ -> calculate_distance(one, other)
+  }
+}
+
+fn calculate_distance(one: String, other: String) -> Int {
+  let n = string.length(one)
+  let m = string.length(other)
+
+  // Convert both strings to dictionaries for key-value access
+  let one_dict = string_to_dict(one)
+  let other_dict = string_to_dict(other)
+
+  // Initialize a matrix to store distance calculations
+  let distance_dict = init_dict(n, m)
+
+  // Compute the final distances between substrings
+  let distance_dict =
+    compute_distances(one_dict, other_dict, distance_dict, n, m)
+
+  // Return the final distance between the full strings
+  let assert Ok(result) = dict.get(distance_dict, #(n, m))
+  result
+}
+
+fn string_to_dict(str: String) -> dict.Dict(Int, String) {
+  // Convert the string to a list of strings and build a dictionary
+  let string_graphemes = string.to_graphemes(str)
+  let string_dict =
+    list.index_fold(string_graphemes, [], fn(acc, item, i) {
+      [#(i, item), ..acc]
+    })
+
+  // Create a dictionary where keys are indices and values are strings
+  dict.from_list(string_dict)
+}
+
+fn init_dict(n: Int, m: Int) -> dict.Dict(#(Int, Int), Int) {
+  let matrix =
+    list.fold(list.range(0, n), [], fn(outer_acc, i) {
+      list.fold(list.range(0, m), outer_acc, fn(inner_acc, j) {
+        // Base cases: Filling the first row and column with distances
+        // corresponding to the insertion and deletion operations
+        case i == 0, j == 0 {
+          True, _ -> {
+            [#(#(i, j), j), ..inner_acc]
+          }
+          _, True -> {
+            [#(#(i, j), i), ..inner_acc]
+          }
+          // Initialize the rest of the matrix with 0.0
+          False, False -> [#(#(i, j), 0), ..inner_acc]
+        }
+      })
+    })
+
+  // Convert the initialized matrix to a dictionary
+  dict.from_list(matrix)
```

Comment:
Same as before, I'd build the dictionary as I go, use `use` and shadow the `acc` variable:
```gleam
use acc, i <- list.fold(list.range(0, n), dict.new())
use acc, j <- list.fold(list.range(0, m), acc)
todo as "insert the values in the dict"
```